### PR TITLE
Constify `Drop` and `Destruct`

### PR DESCRIPTION
**Pull request by compiler-errors**
r? @lcnr or @fee1-dead

### CODE REVIEW

File: tests/ui/traits/const-traits/const-drop-fail.rs [hunk 1]
```diff
@@ -25,6 +25,8 @@ const fn check<T: ~const Destruct>(_: T) {}
 macro_rules! check_all {
     ($($exp:expr),*$(,)?) => {$(
         const _: () = check($exp);
+        //~^ ERROR the trait bound `NonTrivialDrop: const Destruct` is not satisfied
```
Comment (compiler-errors) [inline tests/ui/traits/const-traits/const-drop-fail.rs:28]: The span here sucks because we point at the span of the `check($exp)` expr, rather than pointing at the argument. I plan on fixing that soon.

Comment (compiler-errors) [reply tests/ui/traits/const-traits/const-drop-fail.rs:28]: Soon: https://github.com/rust-lang/rust/pull/133403

File: tests/ui/dropck/const_drop_is_valid.rs [hunk 2]
```diff
@@ -1,8 +1,11 @@
...
+
 struct A();
 
-impl const Drop for A {}
-//~^ ERROR: const trait impls are experimental
-//~| const `impl` for trait `Drop` which is not marked with `#[const_trait]`
-//~| not all trait items implemented, missing: `drop`
+impl const Drop for A {
+    fn drop(&mut self) {}
+}
```
Comment (lcnr) [inline tests/ui/dropck/const_drop_is_valid.rs:9]: that feels like an odd test, why does it exist?

File: tests/ui/traits/const-traits/const-drop.rs [hunk 3]
```diff
@@ -64,7 +65,7 @@ mod t {
 
     pub struct ConstDropWithBound<T: const SomeTrait>(pub core::marker::PhantomData<T>);
 
-    impl<T: ~const SomeTrait> const Drop for ConstDropWithBound<T> {
+    impl<T: const SomeTrait> const Drop for ConstDropWithBound<T> {
```
Comment (lcnr) [inline tests/ui/traits/const-traits/const-drop.rs:68]: why this change?

**Review by lcnr — COMMENTED** [review]
some questions wrt to test changes, otherwise r=me

File: tests/ui/traits/const-traits/const-drop.rs (hunk 3 above)
Comment (compiler-errors) [reply tests/ui/traits/const-traits/const-drop.rs:68]: bc the struct requires unconditional drop

File: tests/ui/dropck/const_drop_is_valid.rs (hunk 2 above)
Comment (compiler-errors) [reply tests/ui/dropck/const_drop_is_valid.rs:9]: sure
